src/query/api/v1/handler/placement/add_test.go

```diff
@@ -37,99 +37,160 @@ import (
+		} else {
+			req = httptest.NewRequest(AddHTTPMethod, M3DBAddURL, strings.NewReader("{\"force\": true, \"instances\":[{\"id\": \"host1\",\"isolation_group\": \"rack1\",\"zone\": \"test\",\"weight\": 1,\"endpoint\": \"http://host1:1234\",\"hostname\": \"host1\",\"port\": 1234}]}"))
+		}
+		require.NotNil(t, req)
+
+		mockPlacementService.EXPECT().AddInstances(gomock.Not(nil)).Return(placement.NewPlacement(), nil, nil)
+		handler.ServeHTTP(serviceName, w, req)
+
+		resp = w.Result()
+		body, _ = ioutil.ReadAll(resp.Body)
+		assert.Equal(t, http.StatusOK, resp.StatusCode)
+		assert.Equal(t, "{\"placement\":{\"instances\":{},\"replicaFactor\":0,\"numShards\":0,\"isSharded\":false,\"cutoverTime\":\"0\",\"isMirrored\":false,\"maxShardSetId\":0},\"version\":0}", string(body))
+
+	})
 }
 
 func TestPlacementAddHandler_SafeErr(t *testing.T) {
-	mockClient, mockPlacementService := SetupPlacementTest(t)
-	handler := NewAddHandler(mockClient, config.Configuration{})
-
-	// Test add failure
-	w := httptest.NewRecorder()
-	req := httptest.NewRequest("POST", "/placement", strings.NewReader("{\"instances\":[]}"))
-	require.NotNil(t, req)
-
-	mockPlacementService.EXPECT().Placement().Return(placement.NewPlacement(), 0, errors.New("no new instances found in the valid zone"))
-	mockPlacementService.EXPECT().AddInstances(gomock.Any()).Return(placement.NewPlacement(), nil, errors.New("no new instances found in the valid zone"))
-	handler.ServeHTTP(w, req)
-
-	resp := w.Result()
-	body, _ := ioutil.ReadAll(resp.Body)
-	assert.Equal(t, http.StatusInternalServerError, resp.StatusCode)
-	assert.Equal(t, "{\"error\":\"no new instances found in the valid zone\"}\n", string(body))
-
-	// Current placement has initializing shards
-	w = httptest.NewRecorder()
-	req = httptest.NewRequest("POST", "/placement", strings.NewReader("{\"instances\":[{\"id\": \"host1\",\"isolation_group\": \"rack1\",\"zone\": \"test\",\"weight\": 1,\"endpoint\": \"http://host1:1234\",\"hostname\": \"host1\",\"port\": 1234}]}"))
-	require.NotNil(t, req)
-
-	mockPlacementService.EXPECT().Placement().Return(newInitPlacement(), 0, nil)
-	mockPlacementService.EXPECT().AddInstances(gomock.Not(nil)).Return(placement.NewPlacement(), nil, nil)
-	handler.ServeHTTP(w, req)
-
-	resp = w.Result()
-	body, _ = ioutil.ReadAll(resp.Body)
-	assert.Equal(t, http.StatusBadRequest, resp.StatusCode)
-	assert.Equal(t, `{"error":"instances [A,B] do not have all shards available"}`+"\n", string(body))
+	runForAllAllowedServices(func(serviceName string) {
+		mockClient, mockPlacementService := SetupPlacementTest(t)
+		handler := NewAddHandler(mockClient, config.Configuration{})
+
+		// Test add failure
+		var (
+			w   = httptest.NewRecorder()
+			req *http.Request
+		)
+		if serviceName == M3AggServiceName {
+			req = httptest.NewRequest(AddHTTPMethod, M3DBAddURL, strings.NewReader("{\"instances\":[], \"max_aggregation_window_size_nanos\": 1000000, \"warmup_duration_nanos\": 1000000}"))
```

Comment:
Do you want to use back ticks to avoid having to escape the quotes?  Looks much more readable, e.g.:
```go
strings.NewReader(`{"instances": [], "max_aggregation_window_size_nanos": 1000000, "warmup_duration_nanos": 1000000}"`)

```